From 0e82ce4b2ef31f94d4f7abda9cb3dd7a3e4aa588 Mon Sep 17 00:00:00 2001
From: Aihua Xu <aihuaxu@apache.org>
Date: Wed, 6 Jan 2016 16:29:54 -0500
Subject: [PATCH 0483/1164] CDH-36225: HIVE-11828: beeline -f fails on scripts
 with tabs between column type and comment (Aihua
 Xu, reviewed by Yongzhi Chen)

Change-Id: I2d8aac33bd59c78240af2f269407d65f280e344e
---
 .../src/java/org/apache/hive/beeline/BeeLine.java  |    1 +
 .../apache/hive/beeline/TestBeeLineWithArgs.java   |    8 ++++++++
 2 files changed, 9 insertions(+)

diff --git a/beeline/src/java/org/apache/hive/beeline/BeeLine.java b/beeline/src/java/org/apache/hive/beeline/BeeLine.java
index 2c129ec..c754ab5 100644
--- a/beeline/src/java/org/apache/hive/beeline/BeeLine.java
+++ b/beeline/src/java/org/apache/hive/beeline/BeeLine.java
@@ -938,6 +938,7 @@ public ConsoleReader getConsoleReader(InputStream inputStream) throws IOExceptio
       InputStream inputStreamAppendedNewline = new SequenceInputStream(inputStream,
           new ByteArrayInputStream((new String("\n")).getBytes()));
       consoleReader = new ConsoleReader(inputStreamAppendedNewline, getOutputStream());
+      consoleReader.setCopyPasteDetection(true); // jline will detect if <tab> is regular character
     } else {
       consoleReader = new ConsoleReader();
     }
diff --git a/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestBeeLineWithArgs.java b/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestBeeLineWithArgs.java
index 90727a7..03861d9 100644
--- a/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestBeeLineWithArgs.java
+++ b/itests/hive-unit/src/test/java/org/apache/hive/beeline/TestBeeLineWithArgs.java
@@ -323,6 +323,14 @@ public void testBreakOnErrorScriptFile() throws Throwable {
   }
 
   @Test
+  public void testTabInScriptFile() throws Throwable {
+    List<String> argList = getBaseArgs(miniHS2.getBaseJdbcURL());
+    final String SCRIPT_TEXT = "CREATE\tTABLE IF NOT EXISTS testTabInScriptFile\n(id\tint);\nSHOW TABLES;";
+    final String EXPECTED_PATTERN = "testTabInScriptFile";
+    testScriptFile(SCRIPT_TEXT, EXPECTED_PATTERN, true, argList);
+  }
+
+  @Test
   public void testBeelineShellCommand() throws Throwable {
     List<String> argList = getBaseArgs(miniHS2.getBaseJdbcURL());
     final String SCRIPT_TEXT = "!sh echo \"hello world.\" > hw.txt\n!sh cat hw.txt\n!rm hw.txt";
-- 
1.7.9.5

